src/renderer/vanilla/util/render.js

```diff
@@ -1,166 +1,51 @@
-const NODE_TYPE_COMMENT = 8;
-const NODE_TYPE_TEXT = 3;
-
 const _cachedTemplateElements = {};
 
-const convertStringToHTML = (templateString) => {
-	const parser = new DOMParser();
-	const document = parser.parseFromString(`<main>${templateString}</main>`, 'text/html');
-
-	return document.querySelector('main');
-};
-
 /**
- * Create an array of the attributes on an element
- * @param  {NamedNodeMap} attributes The attributes on an element
- * @return {Array}                   The attributes on an element as an array of key/value pairs
+ * Parses the given template string and returns a real DOM element.
+ * @param {string} template
+ * @returns {HTMLElement}
  */
-const getAttributes = function (attributes) {
-	return Array.prototype.map.call(attributes, function (attribute) {
-		return {
-			attributeName: attribute.name,
-			value: attribute.value,
-		};
-	});
+const convertStringToHTML = (template) => {
+	const parser = new DOMParser();
+	const document = parser.parseFromString(template.toString(), 'text/html');
+	return document.body;
 };
 
 /**
- * Create a DOM Tree Map for an element
- *
- * @param  {HTMLElement}    element The element to map
- * @param  {Boolean} isSVG   If true, node is within an SVG
- * @return {{ domMap: Array, plainlySetInnerHTML?: boolean, innerHTML?: string }} A DOM tree map
+ * Diff the attributes for a live DOM element against a template DOM element
+ * @param  {Element} templateElement The new template
+ * @param  {Element} domElement The existing DOM node
  */
-const createDOMMap = function (element, isChild = false, isSVG = false, type = 'template') {
-	const children = [...element.childNodes];
-
-	// If the list contains a comment which says "$$plainly-set-inner-html" it means
-	// that the all children of the element will be set as unsafe `innerHTML´
-	// which means that there won't be any diffing or fancy things
-	// but we will just mark it as a pure component and leave it be.
-	for (const child of children) {
-		if (child?.nodeType === NODE_TYPE_COMMENT && child.textContent === '$$plainly-set-inner-html') {
-			return {
-				domMap: [],
-				plainlySetInnerHTML: true,
-				innerHTML: element.innerHTML,
-			};
+const diffAttributes = function (templateElement, domElement) {
+	// remove or update previous attributes
+	for (/** @type {Attr} */ const attribute of Array.from(domElement.attributes)) {
+		if (templateElement.hasAttribute(attribute.name)) {
+			// TODO: handle attributes with "." aka properties...
```

Comment:
@lukas-schardt I think I have seen you use the "dot" notation as attributes correct? I haven't yet implemented this here. I would actually recommend to get rid of this feature entirely. The problem here is that since we are always rendering from a new string, it will always be considered a change because it is still an attribute while diffing. Although the value might not change for subsequent updates, the new template string will have it as an attribute while the actual live DOM element won't have it as an attribute. And therefore the diffing will always have to something and won't be able to return early.

Reply:
Thats right.. I just think the feature is pretty great for passing complex objects or references without having to serialize them. I think that would be a big plus for the renderer. But the way it's build by basically building a dom from two strings that won't be possible. So yeah, we should probably remove it as it would not work as intendend anyway.

Do you have any idea how we could implement passing property references and maybe even functions?

Reply:
I know I would love it very much as well. We only have one string and one real dom. And that is the point. The real dom has lost the (.attribute) information from the string version.
I mean we can keep the feature - but I would write something to the docs that it will slow down the renderer by some degree. But as far as I understand it, it actually doesn't make any difference other than that the attribute will be removed from the dom after diffing. But when sending the html to the client it will still have to be sent... And if so it is just a different way of doing the same thing. And I'm all for there to be only one way of doing things.

I don't have any idea how to implement such functionality in the current form. And much less even for functions...
I think we need to get rid of the string parsing in order to be able to implement it. But if we start using the proposed TemplateParts and others (like lit does) we will loose the simplicity for rendering server side.